tls/s2n_signature_algorithms.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_cipher_suites.h"
+#include "tls/s2n_tls_digest_preferences.h"
+#include "tls/s2n_signature_algorithms.h"
+
+int s2n_set_signature_and_hash_algorithm(struct s2n_connection *conn, struct s2n_map *supported_sig_hash_algs, s2n_hash_algorithm *hash, s2n_signature_algorithm *sig)
+{
+    /* This function could be called in two places: after receiving the
+     * ClientHello and parsing the extensions and cipher suites, and after
+     * receiving the CertificateRequest which includes supported signature/hash
+     * pairs. In both cases, the cipher suite will have been set */
+    s2n_signature_algorithm cipher_suite_sig_alg = conn->secure.cipher_suite->signature_alg;
+    
+    /* Default our signature digest algorithms. For TLS 1.2 this default is different and may be 
+     * overridden by the signature_algorithms extension. If the server chooses an ECDHE_ECDSA 
+     * cipher suite, this will be overridden to SHA1. 
+     */
+    s2n_hash_algorithm hash_alg_chosen = S2N_HASH_MD5_SHA1;
+    s2n_signature_algorithm sig_alg_chosen = S2N_SIGNATURE_RSA;
+    if (cipher_suite_sig_alg == S2N_SIGNATURE_ECDSA) {
+        sig_alg_chosen = S2N_SIGNATURE_ECDSA;
+        hash_alg_chosen = S2N_HASH_SHA1;
```

Comment:
Do we want to use SHA1 as our default? why not SHA256?

Reply:
My interpretation of the RFCs were that this was the required default: https://tools.ietf.org/html/rfc5246#section-7.4.1.4.1

    If the client does not send the signature_algorithms extension, the
    server MUST do the following:

       -  If the negotiated key exchange algorithm is one of (RSA, DHE_RSA,
          DH_RSA, RSA_PSK, ECDH_RSA, ECDHE_RSA), behave as if client had
          sent the value {sha1,rsa}.
       -  If the negotiated key exchange algorithm is one of (DHE_DSS,
          DH_DSS), behave as if the client had sent the value {sha1,dsa}.
       -  If the negotiated key exchange algorithm is one of (ECDH_ECDSA,
          ECDHE_ECDSA), behave as if the client had sent value {sha1,ecdsa}.

    Note: this is a change from TLS 1.1 where there are no explicit
    rules, but as a practical matter one can assume that the peer
    supports MD5 and SHA-1.